// @appbaseio/reactivesearch-vue/dist/es/index.js
import '@appbaseio/reactivecore';
export { componentTypes } from '@appbaseio/reactivecore/lib/utils/constants';
import './_rollupPluginBabelHelpers-ded08042.js';
import 'vue-types';
import '@appbaseio/reactivecore/lib/utils/helper';
import 'emotion';
import '@appbaseio/vue-emotion';
import 'polished';
import './Button-153cd885.js';
import './vueTypes-8986b709.js';
import './Pagination-28e56c5c.js';
export { default as ReactiveList } from './ReactiveList.js';
import '@appbaseio/reactivecore/lib/utils/transform';
import 'redux';
import './index-5509c0bc.js';
import './ComponentWrapper-edaeaceb.js';
import './PreferencesConsumer-86a676c8.js';
import '@vue/babel-helper-vue-jsx-merge-props';
import './Title-863dfa42.js';
import './ListItem-07888a65.js';
import './Card-19ad6d90.js';
import './Flex-7c177a10.js';
import '@appbaseio/reactivecore/lib/actions/analytics';
import 'appbase-js';
import '@appbaseio/analytics';
import 'url-search-params-polyfill';
export { default as ReactiveBase } from './ReactiveBase.js';
import 'hotkeys-js';
import './Tags-7f4665a9.js';
import './CancelSvg-e40552a5.js';
import './Input-c09c0b56.js';
import 'compute-scroll-into-view';
import './Container-d00219f7.js';
import 'vue-highlight-words';
export { default as DataSearch } from './DataSearch.js';
export { R as ReactiveComponentPrivate, S as SearchBox } from './ReactiveComponentPrivate-799f2576.js';
import '@appbaseio/reactivecore/lib/utils/suggestions';
import './FormControlList-99797d0a.js';
import './utils-9e5a16a1.js';
export { default as SingleList } from './SingleList.js';
export { default as MultiList } from './MultiList.js';
import './DropDown-40687515.js';
export { default as SingleDropdownList } from './SingleDropdownList.js';
export { default as MultiDropdownList } from './MultiDropdownList.js';
export { default as ToggleButton } from './ToggleButton.js';
export { default as ReactiveComponent } from './ReactiveComponent.js';
import 'vue-no-ssr';
import './ssr-03607d84.js';
export { default as DynamicRangeSlider } from './DynamicRangeSlider.js';
export { default as SingleRange } from './SingleRange.js';
export { default as MultiRange } from './MultiRange.js';
export { default as RangeSlider } from './RangeSlider.js';
export { default as RangeInput } from './RangeInput.js';
export { default as SelectedFilters } from './SelectedFilters.js';
export { default as ResultCard } from './ResultCard.js';
export { default as ResultList } from './ResultList.js';
export { default as StateProvider } from './StateProvider.js';
import 'gmap-vue';
import 'ngeohash';
export { default as ReactiveGoogleMap } from './ReactiveGoogleMap.js';
import '@appbaseio/reactivecore/lib/reducers/valueReducer';
import '@appbaseio/reactivecore/lib/reducers/queryReducer';
import '@appbaseio/reactivecore/lib/reducers/queryOptionsReducer';
import '@appbaseio/reactivecore/lib/reducers/compositeAggsReducer';
import '@appbaseio/reactivecore/lib/constants';
import '@appbaseio/reactivecore/lib/reducers/dependencyTreeReducer';
import '@appbaseio/reactivecore/lib/utils/graphQL';
import '@appbaseio/reactivecore/lib/actions/utils';
export { default as initReactivesearch } from './initReactivesearch.js';
import version from './version.js';
export { default as version } from './version.js';
import install from './install.js';
export { default as install } from './install.js';

// Add polyfills to support in IE
if (typeof window !== 'undefined' && window.Vue) {
  install(window.Vue);
}
var index = {
  version: version,
  install: install
};

export default index;
