// @appbaseio/reactivesearch-vue/dist/es/SelectedFilters.js
import { Actions, helper } from '@appbaseio/reactivecore';
import '@appbaseio/reactivecore/lib/utils/constants';
import { a as _extends } from './_rollupPluginBabelHelpers-ded08042.js';
import VueTypes from 'vue-types';
import 'emotion';
import '@appbaseio/vue-emotion';
import 'polished';
import { f as filters, B as Button } from './Button-153cd885.js';
import { t as types } from './vueTypes-8986b709.js';
import 'redux';
import { n as decodeHtml, c as connect } from './index-5509c0bc.js';
import { T as Title } from './Title-863dfa42.js';
import { C as Container } from './Container-d00219f7.js';

var setValue = Actions.setValue,
  clearValues = Actions.clearValues,
  resetValuesToDefault = Actions.resetValuesToDefault;
var getClassName = helper.getClassName,
  handleA11yAction = helper.handleA11yAction;
var SelectedFilters = {
  name: 'SelectedFilters',
  props: {
    className: VueTypes.string.def(''),
    clearAllLabel: VueTypes.string.def('Clear All'),
    innerClass: types.style,
    showClearAll: VueTypes.bool.def(true),
    title: types.title,
    resetToDefault: VueTypes.bool.def(false),
    clearAllBlacklistComponents: VueTypes.array,
    resetToValues: VueTypes.object
  },
  inject: {
    theme: {
      from: 'theme_reactivesearch'
    }
  },
  render: function render() {
    var _this = this;
    var h = arguments[0];
    if (this.$scopedSlots["default"]) {
      return this.$scopedSlots["default"]({
        components: this.components,
        selectedValues: this.selectedValues,
        clearValues: this.clearValues,
        clearValue: this.clearValue,
        setValue: this.setValue,
        resetValuesToDefault: this.resetValuesToDefault
      });
    }
    var filtersToRender = this.renderFilters();
    var hasValues = !!filtersToRender.length;
    return h(Container, {
      "class": filters(this.theme) + " " + (this.$props.className || '')
    }, [this.$props.title && hasValues && h(Title, {
      "class": getClassName(this.$props.innerClass, 'title') || ''
    }, [this.$props.title]), filtersToRender, this.$props.showClearAll && hasValues && filtersToRender.length > 1 ? h(Button, {
      "class": getClassName(this.$props.innerClass, 'button') || '',
      "on": _extends({}, {
        click: this.clearValues,
        keypress: function keypress(event) {
          return handleA11yAction(event, function () {
            return _this.clearValues();
          });
        }
      }),
      "attrs": {
        "tabIndex": "0"
      }
    }, [this.$props.clearAllLabel]) : null]);
  },
  methods: {
    remove: function remove(component, value) {
      if (value === void 0) {
        value = null;
      }
      this.setValue(component, null);
      this.$emit('clear', component, value);
    },
    clearValues: function clearValues() {
      var resetToDefault = this.resetToDefault,
        resetToValues = this.resetToValues,
        clearAllBlacklistComponents = this.clearAllBlacklistComponents;
      if (resetToDefault) {
        this.resetValuesToDefault(clearAllBlacklistComponents);
      } else {
        this.clearValuesAction(resetToValues, clearAllBlacklistComponents);
      }
      this.$emit('clear', resetToValues);
    },
    clearValue: function clearValue(componentId) {
      var resetToDefault = this.resetToDefault,
        resetToValues = this.resetToValues;
      if (resetToDefault) {
        this.resetValuesToDefault(this.components.filter(function (component) {
          return component !== componentId;
        }));
      } else {
        this.setValue(componentId, (resetToValues == null ? void 0 : resetToValues[componentId]) || null);
      }
      this.$emit('clear', resetToValues == null ? void 0 : resetToValues[componentId]);
    },
    renderValue: function renderValue(value, isArray) {
      var _this2 = this;
      if (isArray && value.length) {
        var arrayToRender = value.map(function (item) {
          return _this2.renderValue(item);
        });
        return arrayToRender.join(', ');
      }
      if (value && typeof value === 'object') {
        // TODO: support for NestedList
        var label = (typeof value.label === 'string' ? value.label : value.value) || value.key || value.distance || null;
        if (value.location) {
          label = value.location + " - " + label;
        }
        return label;
      }
      return value;
    },
    renderFilterButton: function renderFilterButton(component, keyProp, handleRemove, label) {
      var h = this.$createElement;
      return h(Button, {
        "class": getClassName(this.$props.innerClass, 'button') || '',
        "key": keyProp,
        "on": _extends({}, {
          click: handleRemove,
          keypress: function keypress(event) {
            return handleA11yAction(event, handleRemove);
          }
        }),
        "attrs": {
          "tabIndex": "0"
        }
      }, [h("span", [label]), h("span", ["\u2715"])]);
    },
    renderFilters: function renderFilters() {
      var _this3 = this;
      var selectedValues = this.selectedValues;
      var filterComponents = Object.keys(selectedValues).filter(function (id) {
        return _this3.components.includes(id) && selectedValues[id].showFilter;
      });
      return filterComponents.map(function (component, index) {
        var _selectedValues$compo = selectedValues[component],
          label = _selectedValues$compo.label,
          value = _selectedValues$compo.value;
        var isArray = Array.isArray(value);

        // default behaviour
        if (label && (isArray && value.length || !isArray && value)) {
          var valueToRender = _this3.renderValue(value, isArray);
          return _this3.renderFilterButton(component, component + "-" + (index + 1), function () {
            return _this3.remove(component, value);
          }, selectedValues[component].label + ": " + decodeHtml(valueToRender));
        }
        return null;
      }).filter(Boolean);
    }
  },
  watch: {
    selectedValues: function selectedValues(newVal) {
      this.$emit('change', newVal);
    }
  }
};
var mapStateToProps = function mapStateToProps(state) {
  return {
    components: state.components,
    selectedValues: state.selectedValues
  };
};
var mapDispatchtoProps = {
  clearValuesAction: clearValues,
  setValue: setValue,
  resetValuesToDefault: resetValuesToDefault
};
var RcConnected = connect(mapStateToProps, mapDispatchtoProps)(SelectedFilters);
RcConnected.name = SelectedFilters.name;
RcConnected.install = function (Vue) {
  Vue.component(RcConnected.name, RcConnected);
};

export default RcConnected;
