// @appbaseio/reactivesearch-vue/dist/cjs/RangeInput.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var configureStore = require('@appbaseio/reactivecore');
var configureStore__default = _interopDefault(configureStore);
var constants = require('@appbaseio/reactivecore/lib/utils/constants');
var _rollupPluginBabelHelpers = require('./_rollupPluginBabelHelpers-f8b843f8.js');
var VueTypes = _interopDefault(require('vue-types'));
var emotion = require('emotion');
var styled = require('@appbaseio/vue-emotion');
var styled__default = _interopDefault(styled);
require('polished');
var vueTypes = require('./vueTypes-7cfc2f98.js');
require('@appbaseio/reactivecore/lib/utils/transform');
require('redux');
var index = require('./index-de136c17.js');
var ComponentWrapper = require('./ComponentWrapper-94c50228.js');
var PreferencesConsumer = require('./PreferencesConsumer-5e6e84a2.js');
require('./Title-16042ea0.js');
var Flex = require('./Flex-ddd7cb6b.js');
var Input = require('./Input-ead4cd84.js');
var Container = require('./Container-e699ea95.js');
require('vue-no-ssr');
require('./ssr-f2fc1bf6.js');
var RangeSlider = require('./RangeSlider.js');

var _templateObject, _templateObject2;
var alert = function alert(_ref) {
  var theme = _ref.theme;
  return emotion.css(_templateObject || (_templateObject = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\tcolor: ", ";\n"])), theme.colors.alertColor);
};
var Content = styled__default('div')(_templateObject2 || (_templateObject2 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\t", ";\n\tfont-size: 13px;\n\tmargin: 8px;\n"])), function (props) {
  return props.alert && alert;
});

var getClassName = configureStore.helper.getClassName,
  isEqual = configureStore.helper.isEqual;
var RangeInput = {
  name: 'RangeInput',
  components: {
    RangeSlider: RangeSlider.RangeConnected
  },
  inject: {
    theme: {
      from: 'theme_reactivesearch'
    }
  },
  data: function data() {
    var state = {
      currentValue: {
        start: this.$props.range ? this.$props.range.start : 0,
        end: this.$props.range ? this.$props.range.end : 10
      },
      isStartValid: true,
      isEndValid: true
    };
    return state;
  },
  props: {
    className: {
      types: vueTypes.types.string,
      "default": ''
    },
    defaultValue: vueTypes.types.range,
    validateRange: vueTypes.types.func,
    value: vueTypes.types.range,
    dataField: vueTypes.types.stringRequired,
    innerClass: vueTypes.types.style,
    range: {
      types: vueTypes.types.range,
      "default": function _default() {
        return {
          start: 0,
          end: 10
        };
      }
    },
    rangeLabels: vueTypes.types.rangeLabels,
    stepValue: vueTypes.types.number,
    componentStyle: vueTypes.types.style,
    componentId: vueTypes.types.stringRequired,
    includeNullValues: VueTypes.bool,
    beforeValueChange: vueTypes.types.func,
    customQuery: vueTypes.types.func,
    data: vueTypes.types.data,
    filterLabel: vueTypes.types.string,
    react: vueTypes.types.react,
    showFilter: VueTypes.bool.def(true),
    showCheckbox: VueTypes.bool.def(true),
    title: vueTypes.types.title,
    URLParams: VueTypes.bool.def(false),
    sliderOptions: VueTypes.object.def({}),
    nestedField: vueTypes.types.string,
    endpoint: vueTypes.types.endpointConfig
  },
  methods: {
    shouldUpdate: function shouldUpdate(value) {
      var validateRange = this.$props.validateRange;
      if (validateRange && value) {
        return validateRange([value.start, value.end]);
      }
      if (value) {
        return value.start <= value.end;
      }
      return false;
    },
    isControlled: function isControlled() {
      if (this.$props.value && this.$listeners) {
        return true;
      }
      return false;
    },
    handleChange: function handleChange(value, event) {
      var currentValue = value;
      if (this.shouldUpdate(value) && !isEqual(value, this.currentValue)) {
        switch (event) {
          case 'change':
            if (!value) {
              currentValue = {
                start: this.$props.range ? this.$props.range.start : 0,
                end: this.$props.range ? this.$props.range.end : 10
              };
            }
            this.$data.currentValue = _rollupPluginBabelHelpers._extends({}, currentValue);
            this.$emit('change', this.$data.currentValue);
            break;
          case 'value-change':
            this.$emit('valueChange', this.$data.currentValue);
            this.$emit('value-change', this.$data.currentValue);
            break;
          default:
            this.$data.currentValue = _rollupPluginBabelHelpers._extends({}, currentValue);
            break;
        }
      }
    },
    handleOnChange: function handleOnChange(value) {
      this.handleChange(value || this.$props.range, 'change');
    },
    handleValueChange: function handleValueChange(value) {
      this.handleChange(value, 'value-change');
    },
    handleInputChange: function handleInputChange(e) {
      var _e$target = e.target,
        name = _e$target.name,
        value = _e$target.value;
      if (Number.isNaN(value)) {
        if (name === 'start') {
          this.$data.isStartValid = false;
        } else {
          this.$data.isEndValid = false;
        }
      } else if (name === 'start' && !this.$data.isStartValid) {
        this.$data.isStartValid = true;
      } else if (name === 'end' && !this.$data.isEndValid) {
        this.$data.isEndValid = true;
      }
      if (this.$data.isStartValid && this.$data.isEndValid) {
        if (name === 'start') {
          this.handleChange({
            start: Number(value),
            end: this.$data.currentValue.end
          }, 'change');
        } else {
          this.handleChange({
            start: this.$data.currentValue.start,
            end: Number(value)
          }, 'change');
        }
      }
    }
  },
  watch: {
    defaultValue: function defaultValue(newVal, oldVal) {
      if (oldVal.start !== newVal.start || oldVal.end !== newVal.end) {
        this.handleChange(newVal);
      }
    },
    value: function value(newVal, oldVal) {
      if (!isEqual(newVal, oldVal)) {
        if (this.isControlled()) {
          this.handleChange(newVal, 'change');
        }
      }
    }
  },
  created: function created() {
    if (this.$props.defaultValue && this.$props.defaultValue.start && this.$props.defaultValue.end) {
      this.handleChange(this.$props.defaultValue);
    }
    if (this.isControlled()) {
      this.handleChange(this.$props.value, 'change');
    }
  },
  render: function render() {
    var h = arguments[0];
    var _this$$props = this.$props,
      className = _this$$props.className,
      dataField = _this$$props.dataField,
      range = _this$$props.range,
      rangeLabels = _this$$props.rangeLabels,
      componentId = _this$$props.componentId,
      innerClass = _this$$props.innerClass,
      stepValue = _this$$props.stepValue,
      componentStyle = _this$$props.componentStyle,
      themePreset = _this$$props.themePreset,
      includeNullValues = _this$$props.includeNullValues,
      beforeValueChange = _this$$props.beforeValueChange,
      customQuery = _this$$props.customQuery,
      data = _this$$props.data,
      filterLabel = _this$$props.filterLabel,
      react = _this$$props.react,
      showFilter = _this$$props.showFilter,
      showCheckbox = _this$$props.showCheckbox,
      title = _this$$props.title,
      URLParams = _this$$props.URLParams,
      sliderOptions = _this$$props.sliderOptions,
      nestedField = _this$$props.nestedField;
    return h(Container.Container, {
      "style": componentStyle,
      "class": className
    }, [h(RangeSlider.RangeConnected, {
      "attrs": {
        "componentId": componentId,
        "value": {
          start: this.currentValue.start,
          end: this.currentValue.end
        },
        "range": range,
        "dataField": dataField,
        "rangeLabels": rangeLabels,
        "includeNullValues": includeNullValues,
        "beforeValueChange": beforeValueChange,
        "customQuery": customQuery,
        "data": data,
        "filterLabel": filterLabel,
        "react": react,
        "showFilter": showFilter,
        "showCheckbox": showCheckbox,
        "title": title,
        "uRLParams": URLParams,
        "sliderOptions": sliderOptions,
        "nestedField": nestedField
      },
      "on": {
        "change": this.handleOnChange,
        "value-change": this.handleValueChange
      }
    }), h(Flex.Flex, {
      "class": getClassName(innerClass, 'input-container') || ''
    }, [h(Flex.Flex, {
      "attrs": {
        "direction": "column",
        "flex": 2
      }
    }, [h(Input.Input, {
      "key": componentId + "-start-value",
      "attrs": {
        "name": "start",
        "type": "number",
        "step": stepValue,
        "themePreset": themePreset,
        "aria-label": componentId + "-start-input",
        "min": this.$props.range ? this.$props.range.start : 0,
        "alert": !this.isStartValid
      },
      "on": {
        "change": this.handleInputChange
      },
      "class": getClassName(innerClass, 'input') || '',
      "domProps": _rollupPluginBabelHelpers._extends({}, {
        value: this.currentValue.start
      })
    }), !this.isStartValid && h(Content, {
      "attrs": {
        "alert": true
      }
    }, ["Input range is invalid"])]), h(Flex.Flex, {
      "attrs": {
        "justifyContent": "center",
        "alignItems": "center",
        "flex": 1
      }
    }, ["-"]), h(Flex.Flex, {
      "attrs": {
        "direction": "column",
        "flex": 2
      }
    }, [h(Input.Input, {
      "key": componentId + "-end-value",
      "attrs": {
        "name": "end",
        "type": "number",
        "step": stepValue,
        "themePreset": themePreset,
        "aria-label": componentId + "-end-input",
        "max": this.$props.range ? this.$props.range.end : 10,
        "alert": !this.isEndValid
      },
      "on": {
        "change": this.handleInputChange
      },
      "class": getClassName(innerClass, 'input') || '',
      "domProps": _rollupPluginBabelHelpers._extends({}, {
        value: this.currentValue.end
      })
    }), !this.isEndValid && h(Content, {
      "attrs": {
        "alert": true
      }
    }, ["Input range is invalid"])])])]);
  }
};
var mapStateToProps = function mapStateToProps(state) {
  return {
    themePreset: state.config.themePreset
  };
};
var RangeConnected = PreferencesConsumer.PreferencesConsumer(ComponentWrapper.ComponentWrapper(index.connect(mapStateToProps, {})(RangeInput), {
  componentType: constants.componentTypes.rangeInput
}));
RangeConnected.name = RangeInput.name;
RangeConnected.install = function (Vue) {
  Vue.component(RangeConnected.name, RangeConnected);
};

// Add componentType for SSR
RangeConnected.componentType = constants.componentTypes.rangeInput;

exports.RangeConnected = RangeConnected;
exports.default = RangeConnected;
