// @appbaseio/reactivesearch-vue/dist/cjs/ResultCard.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _rollupPluginBabelHelpers = require('./_rollupPluginBabelHelpers-f8b843f8.js');
var VueTypes = _interopDefault(require('vue-types'));
require('emotion');
require('@appbaseio/vue-emotion');
require('polished');
var vueTypes = require('./vueTypes-7cfc2f98.js');
var _mergeJSXProps = _interopDefault(require('@vue/babel-helper-vue-jsx-merge-props'));
var Title = require('./Title-16042ea0.js');
var Card = require('./Card-240f283d.js');

var ResultCardTitle = {
  name: 'ResultCardTitle',
  render: function render() {
    var h = arguments[0];
    var children = this.$slots["default"];
    return h(Title.Title, _mergeJSXProps([{}, this.$props]), [children]);
  }
};
ResultCardTitle.install = function (Vue) {
  Vue.component(ResultCardTitle.name, ResultCardTitle);
};

var _excluded = ["src"];
var ResultCardImage = {
  name: 'ResultCardImage',
  props: {
    src: VueTypes.string.isRequired
  },
  render: function render() {
    var h = arguments[0];
    var _this$$props = this.$props,
      src = _this$$props.src,
      props = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_this$$props, _excluded);
    return h(Card.Image, _mergeJSXProps([{
      "style": {
        backgroundImage: "url(" + src + ")"
      }
    }, props]));
  }
};
ResultCardImage.install = function (Vue) {
  Vue.component(ResultCardImage.name, ResultCardImage);
};

var ResultCardDescription = {
  name: 'ResultCardDescription',
  functional: true,
  render: function render(_, _ref) {
    var props = _ref.props,
      children = _ref.children;
    var h = arguments[0];
    return h("article", _mergeJSXProps([{}, props]), [children]);
  }
};
ResultCardDescription.install = function (Vue) {
  Vue.component(ResultCardDescription.name, ResultCardDescription);
};

var _excluded$1 = ["href", "target"];
var ResultCard = {
  name: 'ResultCard',
  components: {
    ResultCardTitle: ResultCardTitle,
    ResultCardImage: ResultCardImage,
    ResultCardDescription: ResultCardDescription
  },
  props: {
    target: VueTypes.string.def('_blank'),
    href: vueTypes.types.string
  },
  render: function render() {
    var h = arguments[0];
    var children = this.$slots["default"];
    var _this$$props = this.$props,
      href = _this$$props.href,
      target = _this$$props.target,
      rest = _rollupPluginBabelHelpers._objectWithoutPropertiesLoose(_this$$props, _excluded$1);
    return h(Card.Card, _mergeJSXProps([{
      "attrs": {
        "href": href,
        "target": target,
        "rel": target === '_blank' ? 'noopener noreferrer' : null
      }
    }, rest]), [children]);
  }
};
ResultCard.install = function (Vue) {
  Vue.component(ResultCard.name, ResultCard);
  Vue.component(ResultCardTitle.name, ResultCardTitle);
  Vue.component(ResultCardImage.name, ResultCardImage);
  Vue.component(ResultCardDescription.name, ResultCardDescription);
};

exports.default = ResultCard;
