// @appbaseio/reactivesearch-vue/dist/cjs/Tags-8fbad58b.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var configureStore = require('@appbaseio/reactivecore');
var configureStore__default = _interopDefault(configureStore);
var _rollupPluginBabelHelpers = require('./_rollupPluginBabelHelpers-f8b843f8.js');
var VueTypes = _interopDefault(require('vue-types'));
var styled = require('@appbaseio/vue-emotion');
var styled__default = _interopDefault(styled);
var vueTypes = require('./vueTypes-7cfc2f98.js');
var index = require('./index-de136c17.js');
var Flex = require('./Flex-ddd7cb6b.js');
var CancelSvg = require('./CancelSvg-220a7cc3.js');
var Input = require('./Input-ead4cd84.js');
var Highlight = _interopDefault(require('vue-highlight-words'));

var _templateObject;
var InputGroup = styled__default('div')(_templateObject || (_templateObject = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n  display: flex;\n  align-items: stretch;\n  width: 100%;\n"])));
InputGroup.defaultProps = {
  className: 'input-group'
};

var _templateObject$1;
var InputAddon = styled__default('span')(_templateObject$1 || (_templateObject$1 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n  display: flex;\n  align-items: center;\n  justify-content: center;\n  background-color: #fafafa;\n  border: 1px solid #ccc;\n  border-radius: 2px;\n  color: rgba(0, 0, 0, 0.85);\n  font-size: 14px;\n  font-weight: 400;\n  padding: 0 11px;\n  position: relative;\n  transition: all 0.3s;\n  box-sizing: border-box;\n  overflow: hidden;\n\n  &:first-of-type {\n    border-right: none;\n  }\n  &:last-of-type {\n    border-left: none;\n  }\n"])));
InputAddon.defaultProps = {
  className: 'input-addon'
};

var getClassName = configureStore.helper.getClassName;
var SuggestionWrapper = {
  name: 'SuggestionWrapper',
  props: {
    innerClassName: vueTypes.types.string,
    themePreset: vueTypes.types.themePreset,
    innerClass: vueTypes.types.style
  },
  render: function render() {
    var h = arguments[0];
    var _this$$props = this.$props,
      themePreset = _this$$props.themePreset,
      innerClass = _this$$props.innerClass,
      innerClassName = _this$$props.innerClassName;
    return h("div", {
      "class": Input.noSuggestions(themePreset) + " " + getClassName(innerClass, innerClassName || '')
    }, [h("li", [this.$scopedSlots["default"]()])]);
  }
};

var _templateObject$2;
var highlightStyle = {
  fontWeight: 600,
  padding: 0,
  backgroundColor: 'transparent',
  color: 'inherit'
};
var PredictiveSuggestion = styled__default('span')(_templateObject$2 || (_templateObject$2 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\t.highlight {\n\t\tbackground: transparent;\n\t\tcolor: inherit;\n\t\tfont-weight: 600;\n\t\tpadding: 0;\n\t}\n"])));
var SuggestionItem = {
  name: 'SuggestionItem',
  props: {
    currentValue: vueTypes.types.string,
    suggestion: vueTypes.types.any
  },
  render: function render() {
    var h = arguments[0];
    var _this$suggestion = this.suggestion,
      label = _this$suggestion.label,
      value = _this$suggestion.value,
      title = _this$suggestion.title,
      description = _this$suggestion.description,
      image = _this$suggestion.image,
      isPredictiveSuggestion = _this$suggestion.isPredictiveSuggestion,
      _suggestion_type = _this$suggestion._suggestion_type,
      _category = _this$suggestion._category;
    if (label) {
      // label has highest precedence
      return typeof label === 'string' ? h("div", {
        "class": "trim"
      }, [(_category ? false : isPredictiveSuggestion
      // eslint-disable-next-line
      || !!_suggestion_type) ? h(PredictiveSuggestion, {
        "domProps": {
          "innerHTML": label
        }
      }) : h(Highlight, {
        "attrs": {
          "searchWords": _category ? [_category] : this.currentValue.split(' '),
          "textToHighlight": label,
          "autoEscape": true,
          "highlightStyle": highlightStyle
        }
      })]) : label;
    }
    if (title || image || description) {
      return h(Flex.Flex, {
        "attrs": {
          "alignItems": "center",
          "css": {
            width: '100%'
          }
        }
      }, [image && h("div", {
        "attrs": {
          "css": {
            margin: 'auto',
            marginRight: 10
          }
        }
      }, [h("img", {
        "attrs": {
          "src": image,
          "alt": " ",
          "height": "50px",
          "width": "50px",
          "css": {
            objectFit: 'contain'
          }
        }
      })]), h(Flex.Flex, {
        "attrs": {
          "direction": "column",
          "css": {
            width: image ? 'calc(100% - 60px)' : '100%'
          }
        }
      }, [title && h("div", {
        "class": "trim"
      }, [h(Highlight, {
        "attrs": {
          "searchWords": this.currentValue.split(' '),
          "textToHighlight": title,
          "highlightStyle": highlightStyle
        },
        "class": styled.css({
          fontSize: '1rem'
        })
      })]), description && h("div", {
        "class": styled.cx('trim', styled.css({
          marginTop: 3
        }))
      }, [h(Highlight, {
        "attrs": {
          "searchWords": this.currentValue.split(' '),
          "textToHighlight": description,
          "highlightStyle": highlightStyle
        }
      })])])]);
    }
    return value;
  }
};

var SearchSvg = {
  name: 'SearchSvg',
  props: ['styles'],
  render: function render() {
    var h = arguments[0];
    return h("svg", {
      "attrs": {
        "alt": "Search",
        "height": "12",
        "xmlns": "http://www.w3.org/2000/svg",
        "viewBox": "0 0 15 15"
      },
      "class": "search-icon",
      "style": _rollupPluginBabelHelpers._extends({
        transform: 'scale(1.35)',
        position: 'relative'
      }, this.$props.styles ? this.$props.styles : {})
    }, [h("title", ["Search"]), h("path", {
      "attrs": {
        "d": 'M6.02945,10.20327a4.17382,4.17382,0,1,1,4.17382-4.17382A4.15609,4.15609,0,0,1,6.02945,10.20327Zm9.69195,4.2199L10.8989,9.59979A5.88021,5.88021,0,0,0,12.058,6.02856,6.00467,6.00467,0,1,0,9.59979,10.8989l4.82338,4.82338a.89729.89729,0,0,0,1.29912,0,.89749.89749,0,0,0-.00087-1.29909Z'
      }
    })]);
  }
};

var _templateObject$3;

// eslint-disable-next-line
styled.injectGlobal(_templateObject$3 || (_templateObject$3 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\t#el_TvxDfTAtKp {\n\t\tstroke: none;\n\t\tstroke-width: 1;\n\t\tfill: none;\n\t}\n\t#el_D93PK3GbmJ {\n\t\t-webkit-transform: translate(163px, 131px);\n\t\ttransform: translate(163px, 131px);\n\t\tfill: #d8d8d8;\n\t}\n"])));
var MicSvg = {
  name: 'MicSvg',
  props: ['className'],
  render: function render() {
    var h = arguments[0];
    return h("svg", {
      "attrs": {
        "viewBox": "0 0 480 480",
        "xmlns": "http://www.w3.org/2000/svg",
        "id": "el_xS0FRzQjJ",
        "width": 28,
        "height": 28,
        "className": this.$props.className
      },
      "style": {
        transform: 'scale(1.5)'
      }
    }, [h("g", {
      "attrs": {
        "id": "el_TvxDfTAtKp",
        "fillRule": "evenodd"
      }
    }, [h("g", {
      "attrs": {
        "id": "el_D93PK3GbmJ",
        "fillRule": "nonzero"
      },
      "style": {
        fill: '#595959'
      }
    }, [h("path", {
      "attrs": {
        "d": "M142.731204,111 C137.280427,111 132.719573,114.852 131.82965,120.095 C127.268796,145.24 104.464526,164.5 76.9881611,164.5 C49.5117965,164.5 26.7075263,145.24 22.1466723,120.095 C21.2567496,114.852 16.6958955,111 11.2451187,111 C4.45945784,111 -0.880078594,116.778 0.121084488,123.198 C5.57186127,155.298 32.2695435,180.443 65.8641269,185.044 L65.8641269,207.3 C65.8641269,213.185 70.8699423,218 76.9881611,218 C83.10638,218 88.1121954,213.185 88.1121954,207.3 L88.1121954,185.044 C121.706779,180.443 148.404461,155.298 153.855238,123.198 C154.967641,116.778 149.516864,111 142.731204,111 Z",
        "id": "el_uly3EwA2O3"
      }
    }), h("path", {
      "attrs": {
        "d": "M76.9864699,147.789474 C98.090352,147.789474 115.126016,131.286316 115.126016,110.842105 L115.126016,36.9473684 C115.126016,16.5031579 98.090352,-2.84217094e-14 76.9864699,-2.84217094e-14 C55.8825877,-2.84217094e-14 38.8469239,16.5031579 38.8469239,36.9473684 L38.8469239,110.842105 C38.8469239,131.286316 55.8825877,147.789474 76.9864699,147.789474 Z",
        "id": "el_tnDbR4ytu4"
      }
    })])])]);
  }
};

var _templateObject$4;

// eslint-disable-next-line
styled.injectGlobal(_templateObject$4 || (_templateObject$4 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\t#el_X81iT9kZYo {\n\t\tstroke: none;\n\t\tstroke-width: 1;\n\t\tfill: none;\n\t}\n\t#el_gMpyalCphp {\n\t\t-webkit-transform: translate(163px, 131px);\n\t\ttransform: translate(163px, 131px);\n\t}\n\t#el_c7H-3u-D4l {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_qhFcdAAFwo {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_M8X8g37WOI {\n\t\tstroke: #e83137;\n\t\tstroke-width: 21;\n\t}\n"])));
var MuteSvg = {
  name: 'MuteSvg',
  props: ['className'],
  render: function render() {
    var h = arguments[0];
    return h("svg", {
      "style": {
        transform: 'scale(1.5)'
      },
      "attrs": {
        "viewBox": "0 0 480 480",
        "xmlns": "http://www.w3.org/2000/svg",
        "id": "el_D1rEpH2zj",
        "width": 28,
        "height": 28,
        "className": this.$props.className
      }
    }, [h("g", {
      "attrs": {
        "id": "el_X81iT9kZYo",
        "fillRule": "evenodd"
      }
    }, [h("g", {
      "attrs": {
        "id": "el_gMpyalCphp"
      }
    }, [h("path", {
      "attrs": {
        "d": "M142.731204,111 C137.280427,111 132.719573,114.852 131.82965,120.095 C127.268796,145.24 104.464526,164.5 76.9881611,164.5 C49.5117965,164.5 26.7075263,145.24 22.1466723,120.095 C21.2567496,114.852 16.6958955,111 11.2451187,111 C4.45945784,111 -0.880078594,116.778 0.121084488,123.198 C5.57186127,155.298 32.2695435,180.443 65.8641269,185.044 L65.8641269,207.3 C65.8641269,213.185 70.8699423,218 76.9881611,218 C83.10638,218 88.1121954,213.185 88.1121954,207.3 L88.1121954,185.044 C121.706779,180.443 148.404461,155.298 153.855238,123.198 C154.967641,116.778 149.516864,111 142.731204,111 Z",
        "id": "el_c7H-3u-D4l",
        "fillRule": "nonzero"
      },
      "style": {
        fill: '#595959'
      }
    }), h("path", {
      "attrs": {
        "d": "M76.9864699,147.789474 C98.090352,147.789474 115.126016,131.286316 115.126016,110.842105 L115.126016,36.9473684 C115.126016,16.5031579 98.090352,-2.84217094e-14 76.9864699,-2.84217094e-14 C55.8825877,-2.84217094e-14 38.8469239,16.5031579 38.8469239,36.9473684 L38.8469239,110.842105 C38.8469239,131.286316 55.8825877,147.789474 76.9864699,147.789474 Z",
        "id": "el_qhFcdAAFwo",
        "fillRule": "nonzero"
      },
      "style": {
        fill: '#595959'
      }
    }), h("path", {
      "attrs": {
        "d": "M11.5,206.5 L142.5,12.5",
        "id": "el_M8X8g37WOI",
        "strokeLinecap": "round",
        "strokeLinejoin": "round"
      }
    })])])]);
  }
};

var _templateObject$5;

// eslint-disable-next-line
styled.injectGlobal(_templateObject$5 || (_templateObject$5 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\t@-webkit-keyframes kf_el_6WKby7wXqV_an_qqO-rxbNc {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t13.89% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@keyframes kf_el_6WKby7wXqV_an_qqO-rxbNc {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t13.89% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@-webkit-keyframes kf_el_Wi-my975tM_an_XhXP1epXB {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t27.78% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@keyframes kf_el_Wi-my975tM_an_XhXP1epXB {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t27.78% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@-webkit-keyframes kf_el_DkfFFTaFxy8_an_T2XxzvIaA {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t41.67% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@keyframes kf_el_DkfFFTaFxy8_an_T2XxzvIaA {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t41.67% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@-webkit-keyframes kf_el_34IgwiMB5rf_an_TPom3H2LI {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t55.56% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@keyframes kf_el_34IgwiMB5rf_an_TPom3H2LI {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t55.56% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@-webkit-keyframes kf_el_DeebuCsPTGA_an_aYTRBE7Na {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t69.44% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@keyframes kf_el_DeebuCsPTGA_an_aYTRBE7Na {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t69.44% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@-webkit-keyframes kf_el_ZOjjrPTvyrv_an_l_BjBNzXw {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t83.33% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@keyframes kf_el_ZOjjrPTvyrv_an_l_BjBNzXw {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t83.33% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@-webkit-keyframes kf_el_2FATegVmf0K_an_wLg4ofuFx {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t97.22% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t@keyframes kf_el_2FATegVmf0K_an_wLg4ofuFx {\n\t\t0% {\n\t\t\topacity: 0;\n\t\t}\n\t\t97.22% {\n\t\t\topacity: 1;\n\t\t}\n\t\t100% {\n\t\t\topacity: 1;\n\t\t}\n\t}\n\t#el_hiibMG0x- * {\n\t\t-webkit-animation-duration: 1.2s;\n\t\tanimation-duration: 1.2s;\n\t\t-webkit-animation-iteration-count: infinite;\n\t\tanimation-iteration-count: infinite;\n\t\t-webkit-animation-timing-function: cubic-bezier(0, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0, 0, 1, 1);\n\t}\n\t#el_QJeJ_2CDw5 {\n\t\tstroke: none;\n\t\tstroke-width: 1;\n\t\tfill: none;\n\t}\n\t#el_UYYCfubTRf {\n\t\t-webkit-transform: translate(163px, 123px);\n\t\ttransform: translate(163px, 123px);\n\t}\n\t#el_uzZNtK32Zi {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_EYKQ2N9Kgy {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_6SDP2LAgKC {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t}\n\t#el_-Vm65Ltfy7 {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_q04iZcSim4 {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_6WKby7wXqV {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t\t-webkit-animation-fill-mode: backwards;\n\t\tanimation-fill-mode: backwards;\n\t\topacity: 0;\n\t\t-webkit-animation-name: kf_el_6WKby7wXqV_an_qqO-rxbNc;\n\t\tanimation-name: kf_el_6WKby7wXqV_an_qqO-rxbNc;\n\t\t-webkit-animation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t}\n\t#el_9bggsfQOtU {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_NKxqi9eIym {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_Wi-my975tM {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t\t-webkit-animation-fill-mode: backwards;\n\t\tanimation-fill-mode: backwards;\n\t\topacity: 0;\n\t\t-webkit-animation-name: kf_el_Wi-my975tM_an_XhXP1epXB;\n\t\tanimation-name: kf_el_Wi-my975tM_an_XhXP1epXB;\n\t\t-webkit-animation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t}\n\t#el_zclQ34fvf7 {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_1OsvRT8HkeZ {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_DkfFFTaFxy8 {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t\t-webkit-animation-fill-mode: backwards;\n\t\tanimation-fill-mode: backwards;\n\t\topacity: 0;\n\t\t-webkit-animation-name: kf_el_DkfFFTaFxy8_an_T2XxzvIaA;\n\t\tanimation-name: kf_el_DkfFFTaFxy8_an_T2XxzvIaA;\n\t\t-webkit-animation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t}\n\t#el_aa9sjx4H0vA {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_tea114vWg0J {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_34IgwiMB5rf {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t\t-webkit-animation-fill-mode: backwards;\n\t\tanimation-fill-mode: backwards;\n\t\topacity: 0;\n\t\t-webkit-animation-name: kf_el_34IgwiMB5rf_an_TPom3H2LI;\n\t\tanimation-name: kf_el_34IgwiMB5rf_an_TPom3H2LI;\n\t\t-webkit-animation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t}\n\t#el_z5u6RAFhx7d {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_7nfuWmA5Uhy {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_DeebuCsPTGA {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t\t-webkit-animation-fill-mode: backwards;\n\t\tanimation-fill-mode: backwards;\n\t\topacity: 0;\n\t\t-webkit-animation-name: kf_el_DeebuCsPTGA_an_aYTRBE7Na;\n\t\tanimation-name: kf_el_DeebuCsPTGA_an_aYTRBE7Na;\n\t\t-webkit-animation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t}\n\t#el__ZcqlS20zcw {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_8DnEQnD7VWV {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_ZOjjrPTvyrv {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t\t-webkit-animation-fill-mode: backwards;\n\t\tanimation-fill-mode: backwards;\n\t\topacity: 0;\n\t\t-webkit-animation-name: kf_el_ZOjjrPTvyrv_an_l_BjBNzXw;\n\t\tanimation-name: kf_el_ZOjjrPTvyrv_an_l_BjBNzXw;\n\t\t-webkit-animation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t}\n\t#el_FYYKCI_u24e {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_XZty4MnTp5Y {\n\t\tfill: #d8d8d8;\n\t}\n\t#el_2FATegVmf0K {\n\t\t-webkit-transform: translate(37.846924px, 0px);\n\t\ttransform: translate(37.846924px, 0px);\n\t\t-webkit-animation-fill-mode: backwards;\n\t\tanimation-fill-mode: backwards;\n\t\topacity: 0;\n\t\t-webkit-animation-name: kf_el_2FATegVmf0K_an_wLg4ofuFx;\n\t\tanimation-name: kf_el_2FATegVmf0K_an_wLg4ofuFx;\n\t\t-webkit-animation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t\tanimation-timing-function: cubic-bezier(0.42, 0, 1, 1);\n\t}\n\t#el_RMT1KUfbdF8 {\n\t\tfill: #0B6AFF;\n\t}\n\t#el_RgLcovvFiO1 {\n\t\tfill: #d8d8d8;\n\t}\n"])));
var ListenSvg = {
  name: 'ListenSvg',
  props: ['className'],
  render: function render() {
    var h = arguments[0];
    return h("svg", {
      "attrs": {
        "viewBox": "0 0 480 480",
        "xmlns": "http://www.w3.org/2000/svg",
        "xmlnsXlink": "http://www.w3.org/1999/xlink",
        "id": "el_hiibMG0x-",
        "width": 28,
        "height": 29,
        "className": this.$props.className
      },
      "style": {
        transform: 'scale(1.5)'
      }
    }, [h("defs", [h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-1"
      }
    }), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-3"
      }
    }), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-5"
      }
    }), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-7"
      }
    }), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-9"
      }
    }), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-11"
      }
    }), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-13"
      }
    }), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "path-15"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_QJeJ_2CDw5",
        "fillRule": "evenodd"
      }
    }, [h("g", {
      "attrs": {
        "id": "el_UYYCfubTRf"
      }
    }, [h("path", {
      "attrs": {
        "d": "M142.731204,111 C137.280427,111 132.719573,114.852 131.82965,120.095 C127.268796,145.24 104.464526,164.5 76.9881611,164.5 C49.5117965,164.5 26.7075263,145.24 22.1466723,120.095 C21.2567496,114.852 16.6958955,111 11.2451187,111 C4.45945784,111 -0.880078594,116.778 0.121084488,123.198 C5.57186127,155.298 32.2695435,180.443 65.8641269,185.044 L65.8641269,207.3 C65.8641269,213.185 70.8699423,218 76.9881611,218 C83.10638,218 88.1121954,213.185 88.1121954,207.3 L88.1121954,185.044 C121.706779,180.443 148.404461,155.298 153.855238,123.198 C154.967641,116.778 149.516864,111 142.731204,111 Z",
        "id": "el_uzZNtK32Zi",
        "fillRule": "nonzero"
      },
      "style": {
        fill: '#0B6AFF'
      }
    }), h("path", {
      "attrs": {
        "d": "M76.9864699,147.789474 C98.090352,147.789474 115.126016,131.286316 115.126016,110.842105 L115.126016,36.9473684 C115.126016,16.5031579 98.090352,0 76.9864699,0 C55.8825877,0 38.8469239,16.5031579 38.8469239,36.9473684 L38.8469239,110.842105 C38.8469239,131.286316 55.8825877,147.789474 76.9864699,147.789474 Z",
        "id": "el_EYKQ2N9Kgy",
        "fillRule": "nonzero"
      }
    }), h("g", {
      "attrs": {
        "id": "el_6SDP2LAgKC"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-2",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-1"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el_-Vm65Ltfy7",
        "fillRule": "nonzero",
        "mask": "url(#mask-2)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_q04iZcSim4",
        "mask": "url(#mask-2)",
        "x": "0.279",
        "width": "77",
        "height": "130"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_6WKby7wXqV"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-4",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-3"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el_9bggsfQOtU",
        "fillRule": "nonzero",
        "mask": "url(#mask-4)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_NKxqi9eIym",
        "mask": "url(#mask-4)",
        "x": "0.279",
        "width": "77",
        "height": "115"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_Wi-my975tM"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-6",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-5"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el_zclQ34fvf7",
        "fillRule": "nonzero",
        "mask": "url(#mask-6)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_1OsvRT8HkeZ",
        "mask": "url(#mask-6)",
        "x": "0.279",
        "width": "77",
        "height": "100"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_DkfFFTaFxy8"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-8",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-7"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el_aa9sjx4H0vA",
        "fillRule": "nonzero",
        "mask": "url(#mask-8)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_tea114vWg0J",
        "mask": "url(#mask-8)",
        "x": "0.279",
        "width": "77",
        "height": "85"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_34IgwiMB5rf"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-10",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-9"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el_z5u6RAFhx7d",
        "fillRule": "nonzero",
        "mask": "url(#mask-10)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_7nfuWmA5Uhy",
        "mask": "url(#mask-10)",
        "x": "0.279",
        "width": "77",
        "height": "70"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_DeebuCsPTGA"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-12",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-11"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el__ZcqlS20zcw",
        "fillRule": "nonzero",
        "mask": "url(#mask-12)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_8DnEQnD7VWV",
        "mask": "url(#mask-12)",
        "x": "0.279",
        "width": "77",
        "height": "55"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_ZOjjrPTvyrv"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-14",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-13"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el_FYYKCI_u24e",
        "fillRule": "nonzero",
        "mask": "url(#mask-14)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_XZty4MnTp5Y",
        "mask": "url(#mask-14)",
        "x": "0.279",
        "width": "77",
        "height": "40"
      }
    })]), h("g", {
      "attrs": {
        "id": "el_2FATegVmf0K"
      }
    }, [h("mask", {
      "attrs": {
        "id": "mask-16",
        "fill": "#fff"
      }
    }, [h("use", {
      "attrs": {
        "xlink:href": "#path-15"
      }
    })]), h("path", {
      "attrs": {
        "d": "M38.779092,147.789474 C60.0824253,147.789474 77.279092,131.286316 77.279092,110.842105 L77.279092,36.9473684 C77.279092,16.5031579 60.0824253,0 38.779092,0 C17.4757586,0 0.279091964,16.5031579 0.279091964,36.9473684 L0.279091964,110.842105 C0.279091964,131.286316 17.4757586,147.789474 38.779092,147.789474 Z",
        "id": "el_RMT1KUfbdF8",
        "fillRule": "nonzero",
        "mask": "url(#mask-16)"
      }
    }), h("rect", {
      "attrs": {
        "id": "el_RgLcovvFiO1",
        "mask": "url(#mask-16)",
        "x": "0.279",
        "width": "77",
        "height": "25"
      }
    })])])])]);
  }
};

var STATUS = {
  inactive: 'INACTIVE',
  stopped: 'STOPPED',
  active: 'ACTIVE',
  denied: 'DENIED'
};
var Mic = {
  name: 'Mic',
  props: {
    children: vueTypes.types.title,
    lang: VueTypes.string.def('en-US'),
    iconPosition: VueTypes.string.def('left'),
    handleResult: vueTypes.types.func,
    onNoMatch: vueTypes.types.func,
    onError: vueTypes.types.func,
    getInstance: vueTypes.types.func,
    render: vueTypes.types.func,
    className: vueTypes.types.string,
    applyClearStyle: VueTypes.bool,
    showIcon: VueTypes.bool
  },
  methods: {
    getComponent: function getComponent() {
      var status = this.$data.status;
      var data = {
        handleClick: this.handleClick,
        status: status
      };
      return index.getComponent(data, this);
    },
    stopMic: function stopMic() {
      if (this.instance) {
        this.status = STATUS.inactive;
        this.instance.stop();
        this.instance = null;
      }
    },
    handleClick: function handleClick() {
      var _this = this;
      this.results = [];
      if (window && window.SpeechRecognition) {
        var status = this.$data.status;
        if (status === STATUS.active) {
          this.status = STATUS.inactive;
        }
        var _this$$props = this.$props,
          handleResult = _this$$props.handleResult,
          onNoMatch = _this$$props.onNoMatch,
          onError = _this$$props.onError,
          lang = _this$$props.lang,
          getInstance = _this$$props.getInstance;
        var _window = window,
          SpeechRecognition = _window.SpeechRecognition;
        if (this.instance && this.status !== STATUS.denied) {
          this.status = STATUS.inactive;
          this.instance.stop();
          this.instance = null;
          return;
        }
        this.instance = new SpeechRecognition();
        this.instance.continuous = true;
        this.instance.interimResults = true;
        this.instance.lang = lang;
        if (getInstance) {
          getInstance(this.instance);
        }
        this.instance.start();
        this.instance.onstart = function () {
          _this.status = STATUS.active;
        };
        this.instance.onresult = function (_ref) {
          var results = _ref.results,
            timeStamp = _ref.timeStamp;
          _this.stopMic();
          if (handleResult) {
            handleResult({
              results: results,
              timeStamp: timeStamp
            });
          }
          _this.results.push({
            results: results,
            timeStamp: timeStamp
          });
        };
        this.instance.onnomatch = function (e) {
          return onNoMatch ? onNoMatch(e) : console.warn(e);
        };
        this.instance.onerror = function (e) {
          if (e.error === 'no-speech' || e.error === 'audio-capture') {
            _this.status = STATUS.inactive;
          } else if (e.error === 'not-allowed') {
            _this.status = STATUS.denied;
          }
          console.error(e);
          if (onError) {
            onError(e);
          }
        };

        /* Below Two methods run when Continuous is False */
        this.instance.onspeechend = function () {
          _this.status = STATUS.inactive;
        };
        this.instance.onaudioend = function () {
          _this.status = STATUS.inactive;
        };
      }
    }
  },
  computed: {
    hasCustomRenderer: function hasCustomRenderer() {
      return index.hasCustomRenderer(this);
    },
    Icon: function Icon() {
      var h = this.$createElement;
      var status = this.$data.status;
      var className = this.$props.className;
      switch (status) {
        case STATUS.active:
          return h(ListenSvg, {
            "attrs": {
              "className": className
            },
            "nativeOn": {
              "click": this.handleClick
            }
          });
        case STATUS.stopped:
        case STATUS.denied:
          return h(MuteSvg, {
            "attrs": {
              "className": className
            },
            "nativeOn": {
              "click": this.handleClick
            }
          });
        default:
          return h(MicSvg, {
            "attrs": {
              "className": className
            },
            "nativeOn": {
              "click": this.handleClick
            }
          });
      }
    }
  },
  data: function data() {
    return {
      status: STATUS.inactive
    };
  },
  created: function created() {
    this.results = [];
    if (typeof window !== 'undefined') {
      window.SpeechRecognition = window.webkitSpeechRecognition || window.SpeechRecognition || null;
      if (!window.SpeechRecognition) {
        console.error('SpeechRecognition is not supported in this browser. Please check the browser compatibility at https://developer.mozilla.org/en-US/docs/Web/API/SpeechRecognition#Browser_compatibility.');
      }
    }
  },
  render: function render() {
    var h = arguments[0];
    if (this.hasCustomRenderer) {
      return this.getComponent();
    }
    return h(CancelSvg.IconWrapper, [this.Icon]);
  }
};

var CustomSvg = {
  name: 'CustomSvg',
  props: {
    className: String,
    icon: Function,
    type: String
  },
  data: function data() {
    return {
      customIcon: this.$props.icon && typeof this.$props.icon === 'function' ? this.$props.icon() : null
    };
  },
  render: function render() {
    var h = arguments[0];
    if (this.customIcon) {
      return h("div", {
        "class": this.$props.className
      }, [this.customIcon]);
    }
    if (this.$props.type === 'recent-search-icon') {
      return h("svg", {
        "attrs": {
          "xmlns": "http://www.w3.org/2000/svg",
          "alt": "Recent Search",
          "height": "20",
          "width": "20",
          "viewBox": "0 0 24 24"
        },
        "style": {
          fill: '#707070'
        },
        "class": this.$props.className
      }, [h("path", {
        "attrs": {
          "d": "M0 0h24v24H0z",
          "fill": "none"
        }
      }), h("path", {
        "attrs": {
          "d": "M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
        }
      })]);
    }
    if (this.$props.type === 'promoted-search-icon') {
      return h("svg", {
        "attrs": {
          "xmlns": "http://www.w3.org/2000/svg",
          "width": "20",
          "alt": "promoted search",
          "height": "20",
          "viewBox": "0 0 24 24"
        },
        "class": this.$props.className,
        "style": {
          fill: '#707070'
        }
      }, [h("path", {
        "attrs": {
          "d": "M12 .587l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
        }
      })]);
    }
    if (this.$props.type === 'popular-search-icon') {
      return h("svg", {
        "attrs": {
          "xmlns": "http://www.w3.org/2000/svg",
          "alt": "Popular Search",
          "height": "20",
          "width": "20",
          "viewBox": "0 0 24 24"
        },
        "style": {
          fill: '#707070'
        },
        "class": this.$props.className
      }, [h("path", {
        "attrs": {
          "d": "M0 0h24v24H0z",
          "fill": "none"
        }
      }), h("path", {
        "attrs": {
          "d": "M16 6l2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
        }
      })]);
    }
    return h(SearchSvg, {
      "attrs": {
        "styles": {
          position: 'relative',
          fill: '#707070',
          left: '3px',
          marginRight: '8px'
        }
      }
    });
  }
};

var _templateObject$6, _templateObject2;
var TagsContainer = styled__default('div')(_templateObject$6 || (_templateObject$6 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\tdisplay: flex;\n\tflex-wrap: wrap;\n\tmargin-top: 10px;\n\twidth: 100%;\n"])));
var TagItem = styled__default('span')(_templateObject2 || (_templateObject2 = _rollupPluginBabelHelpers._taggedTemplateLiteralLoose(["\n\tcursor: default;\n\tbox-sizing: border-box;\n\tmargin: 0 4px 4px 0;\n\tcolor: #000000d9;\n\tfont-size: 14px;\n\tfont-variant: tabular-nums;\n\tline-height: 1.5715;\n\tlist-style: none;\n\tfont-feature-settings: 'tnum';\n\tdisplay: inline-block;\n\theight: auto;\n\tpadding: 0 2px 0 7px;\n\tfont-size: 12px;\n\tline-height: 22px;\n\twhite-space: nowrap;\n\tbackground: #fafafa;\n\tborder: 1px solid #d9d9d9;\n\tborder-radius: 2px;\n\topacity: 1;\n\ttransition: all 0.3s;\n\n\tspan.close-icon {\n\t\tsvg {\n\t\t\tcursor: pointer;\n\t\t\theight: 15px;\n\t\t\tposition: relative;\n\t\t\ttop: 4px;\n\t\t\tfill: #262626;\n\t\t}\n\t}\n"])));

exports.CustomSvg = CustomSvg;
exports.InputAddon = InputAddon;
exports.InputGroup = InputGroup;
exports.Mic = Mic;
exports.SearchSvg = SearchSvg;
exports.SuggestionItem = SuggestionItem;
exports.SuggestionWrapper = SuggestionWrapper;
exports.TagItem = TagItem;
exports.TagsContainer = TagsContainer;
